From 900f53d517d99e5239232e06f7cb317d07f44a29 Mon Sep 17 00:00:00 2001
From: Radu Alexe <radu.alexe@nxp.com>
Date: Wed, 8 Nov 2017 17:22:12 +0200
Subject: [PATCH 356/741] crypto: caam - add caam_dma device on caam_probe
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

commit 29d35535a126cc6fe963fb90a73a09d481e98156 from
http://source.codeaurora.org/external/qoriq/qoriq-components/linux

Dynamically create a platform device for the caam_dma driver
at caam_probe() time.

Signed-off-by: Radu Alexe <radu.alexe@nxp.com>
Signed-off-by: Horia GeantÄƒ <horia.geanta@nxp.com>
Signed-off-by: Zhantao Tang <zhantao.tang@windriver.com>
---
 drivers/crypto/caam/ctrl.c | 19 +++++++++++++++++++
 1 file changed, 19 insertions(+)

diff --git a/drivers/crypto/caam/ctrl.c b/drivers/crypto/caam/ctrl.c
index fec39c35c877..1b05e3636ba6 100644
--- a/drivers/crypto/caam/ctrl.c
+++ b/drivers/crypto/caam/ctrl.c
@@ -25,6 +25,8 @@ EXPORT_SYMBOL(caam_dpaa2);
 #include "qi.h"
 #endif
 
+static struct platform_device *caam_dma_dev;
+
 /*
  * i.MX targets tend to have clock control subsystems that can
  * enable/disable clocking to our device.
@@ -339,6 +341,9 @@ static int caam_remove(struct platform_device *pdev)
 	debugfs_remove_recursive(ctrlpriv->dfs_root);
 #endif
 
+	if (caam_dma_dev)
+		platform_device_unregister(caam_dma_dev);
+
 	/* Unmap controller region */
 	iounmap(ctrl);
 
@@ -506,6 +511,10 @@ static int caam_probe(struct platform_device *pdev)
 		{.family = "Freescale i.MX"},
 		{},
 	};
+	static struct platform_device_info caam_dma_pdev_info = {
+		.name = "caam-dma",
+		.id = PLATFORM_DEVID_NONE
+	};
 	struct device *dev;
 	struct device_node *nprop, *np;
 	struct caam_ctrl __iomem *ctrl;
@@ -759,6 +768,16 @@ static int caam_probe(struct platform_device *pdev)
 		goto caam_remove;
 	}
 
+        caam_dma_pdev_info.parent = dev;
+        caam_dma_pdev_info.dma_mask = dma_get_mask(dev);
+        caam_dma_dev = platform_device_register_full(&caam_dma_pdev_info);
+        if (IS_ERR(caam_dma_dev)) {
+                dev_err(dev, "Unable to create and register caam-dma dev\n");
+                caam_dma_dev = 0;
+        } else {
+                set_dma_ops(&caam_dma_dev->dev, get_dma_ops(dev));
+        }
+
 	if (ctrlpriv->era < 10)
 		rng_vid = (rd_reg32(&ctrl->perfmon.cha_id_ls) &
 			   CHA_ID_LS_RNG_MASK) >> CHA_ID_LS_RNG_SHIFT;
-- 
2.17.1

